src/design_notes/variadic_generics.md

```diff
@@ -0,0 +1,288 @@
+
+Proposes both a syntax similar to C++ as well as a Tuple trait that will be implemented by all tuples.
+The trait would contain helpful types and methods for working with variadic tuples:
+
+- AsRefs type, `(A, B, C) -> (&A, &B, &C)`
+- AsMuts type, `(A, B, C) -> (&mut A, &mut B, &mut C)`
+- elements_as_refs fn, with signature `(&'a self) -> Self::AsRefs<'a>`
+- elements_as_mut fn, with signature `(&'a mut self) -> Self::AsMuts<'a>`
+
+- Not provided, but proposed as future extensions:
+  - Allowing unpacking tuple types in an argument position, preventing the need to call variadic functions like
+  `foo((1, 2.0, "3"))`
+  - Allowing the `...` syntax in generic type position, preventing the need to write traits such as `Fn` like
+  `Fn<(T1, T2, T3)>`
+
+#### Pros
+
+- Proposes simple syntax, similar to eddyb proposal
+
+- Includes Tuple trait, which is helpful for working with variadic arguments
+
+#### Cons
+
+- None yet.
+
+#### Syntax
+
+```rust
+trait MyTrait: Tuple {}
+
+impl MyTrait for () {}
+
+impl<Head, Tail> MyTrait for (Head, ...Tail) where Tail: Tuple {}
+
+fn foo<T: MyTrait>(args: T) {}
+```
+
+### [Fredpointzero's proposal](https://github.com/rust-lang/rfcs/pull/2775)
+
+Takes a similar path to previous proposals in terms of syntax, but adds on a lot of ergonomics syntax.
+This additional syntax is proposed to allow easier usage of the variadic types, allowing the user to generate
+more imperative loops over the variadic type. A lot of the debate on this proposal surrounded the proposed for
+loop syntax, as many found it hard to read/comprehend at a glance compared to the rest of the proposed syntax.
+
+#### Pros
+
+- Ergonomics, this proposal allows more linear usage of variadic types, which the other proposals lack.
+
+- Power, this proposal allows almost full control over the bounds on variadics, and provides many example
+  implementations.
+
+#### Cons
+
+- Complexity, much more new syntax is needed than in the other proposals
+
+- No way to terminate recursive functions currently, they just won't compile
+
+#### Syntax
+
+```rust
+struct Foo<(..T)>
+where
+    ..(T: Debug)
+{
+    items: (..T)
+}
+
+fn append<(..L), (..R)>(l: (..L), r: (..R)) -> (..L, ..R) {
+    todo!()
+}
+
+fn foo<(..T)>(args: (..T)) {
+    let (head, tail @ ..) = args;
+
+    (for arg <ARG> @in args <T> {
+        Vec::<Arg>::new();
+    })
+}
+```
+
+### [Jules Bertholet's draft](https://hackmd.io/@x5qEAKM5Q36wKV_Wftw96g/HJFy6uzDh)
+
+Most similar to Fredpointzero's proposal. Like all previous proposals, uses tuples to model variadics. Unlike all other proposals, variadic lifetime and const generics are fully supported.
+
+#### Pros
+
+- Most flexible proposal so far
+  - Lifetime and const variadics
+  - Generalized MxN -> NxM transformation
+  - Homogeneous varargs, using arrays
+
+#### Cons
+
+- Most complex of all proposals so far
+- No support for recursion, out-of-order iteration, or random access as of yet
```

Comment:
This doesn't sound like a true con. Can it easily be extended to support those, or are there obstacles created by the design (or how it interacts with the language) itself?